pep-0544.txt

```diff
@@ -0,0 +1,949 @@
+            self.value = value
+            self.leafs = leafs
+
+    def walk(graph: Traversable) -> None:
+        ...
+    tree: Tree[float] = Tree(0, [])
+    walk(tree) # OK, 'Tree[float]' is a subtype of 'Traversable'
+
+Generic protocol types follow the same rules of variance as non-protocol
+types. Protocols can be used in all special type constructors provided
+by the ``typing`` module and follow the corresponding subtyping rules.
+For example, protocol ``PInt`` is a subtype of ``Union[PInt, int]``.
+
+
+``Union[]`` and ``All[]``
+-------------------------
+
+``Union`` of protocol classes behaves the same way as for non-protocol
+classes. For example::
+
+  from typing import Union, Optional, Protocol
+
+  class Exitable(Protocol):
+      @abstracrmethod
+      def exit(self) -> int:
+          ...
+  class Quitable(Protocol):
+      @abstractmethod
+      def quit(self) -> Optional[int]:
+          ...
+
+  def finish(task: Union[Exitable, Quitable]) -> int:
+      ...
+  class GoodJob:
+      ...
+      def quit(self) -> int:
+          return 0
+  finish(GoodJob()) # OK
+
+In addition, we propose to add another special type construct
+``All`` that represents intersection types. Although for normal types
+it is not very useful, there are many situations where a variable should
+implement more than one protocol. Annotation by ``All[Proto1, Proto2, ...]``
+means that a given variable or parameter must implement all protocols
+``Proto1``, ``Proto2``, etc. either implicitly or explicitly. Example::
+
+  from typing import Sequence, Hashable, All
+
+  def cached_func(args: All[Sequence[float], Hashable]) -> float:
+      ...
+  cached_func((1, 2, 3)) # OK, tuple is hashable and sequence
+
+The interaction between union and intersection types is specified by PEP 483,
+and basically reflects the corresponding interactions for sets.
+
+
+``Type[]`` with protocols
+-------------------------
+
+Variables and parameters annotated with ``Type[Proto]`` accept only concrete
+(non-protocol) subtypes of ``Proto``. For example::
+
+  class Proto(Protocol):
+      @abstractmethod
+      def meth(self) -> int:
+          ...
+  class Concrete:
+      def meth(self) -> int:
+          return 42
+
+  def fun(cls: Type[Proto]) -> int:
+      return cls().meth() # OK
+  fun(Proto)              # Error
+  fun(Concrete)           # OK
+
+The same rule applies to variables::
+
+  var: Type[Proto]
+  var = Proto    # Error
+  var = Concrete # OK
+  var().meth()   # OK
+
+Assigning a protocol class to a variable is allowed if it is not explicitly
+typed, and such assignment creates a type alias. For non-protocol classes,
+the behavior of ``Type[]`` is not changed.
+
+
+``NewType()`` and type aliases
+------------------------------
+
+Protocols are essentially anonymous. To emphasize this point, static type
+checkers might refuse protocol classes inside ``NewType()`` to avoid an
+illusion that a distinct type is provided::
```

Comment:
I agree with @ddfisher, it doesn't fit my brain to have a NewType that is distinct from another Protocol, given that both are abstract and structural (and thus equivalent). To explain this differently:
- `What = NewType('What', SomeProtocol)` doesn't make sense because any concrete implementation of a protocol should both conform to SomeProtocol **and** What.